src/module.c

```diff
@@ -723,7 +738,7 @@ void moduleFreeContext(RedisModuleCtx *ctx) {
          * outside of call() context (timers, events, etc.). */
         if (--server.module_ctx_nesting == 0) {
             if (!server.core_propagates)
```

Comment:
perhaps we can get rid of core_propagates and use
`if (server.in_nested_call == 0)` here and `if (server.module_ctx_nesting == 0)` in the other places it's used

i think it will achieve the same goal (prevent afterDatasetChange from doing stuff while not at the top-most level of command execution)

Reply:
Lets do it on another PR? I want to keep this PR easy to cherrypick.

Reply:
so in that case let's drop afterDatasetChange and call  firePostKeySpaceJobs from propagatePendingCommands

Reply:
i agree, i think we need to modify these `if`s to look at nesting level and other variables, and not on anything specific to propagation.

@MeirShpilraien is your comment about cherry picking because this code is already different in 7.0? or because you want the change to be conservative (not have any side effects if the feature isn't used)?

let's put the time to map the side effects of such a change so that we feel comfortable with it.

Reply:
@oranagra I want it to be conservative so we will not be afraid to cherrypick it, I can apply all your suggestions to unstable on another PR.

Reply:
i'm ok being conservative for a while and do some cleanup later, but for this to be effective, we need to open a big time gap between the merging of the first PR and the second.

in any case, even if we want to be conservative (i.e. avoid risk of any bugs from this change when the new API isn't used), i think we still need the code to make sense. and it's odd to see a call to `afterDatasetChange` that's inside a condition on `core_propagates`. so maybe the middle ground is to call `afterDatasetChange` unconditionally, but pass a boolean argument specifying whether or not it should call `propagatePendingCommands`